### DISCUSSION TITLE
How to ensure find_package finds vcpkg installed package?

### DISCUSSION THREAD

**Discussion by MarkCallow** [microsoft/vcpkg#20972] [Q&A]
For historical reasons I have 2 versions of the OpenImageIO package installed on my system, one by MacPorts in `/opt/local` and the other by vcpkg in `~/github/vcpkg/installed/x64-osx`. I need my CMake project to find and use the vcpkg version but it is finding the MacPorts version. Here is the find_package call from my project's `CMakeLists.txt`:

```
find_package(openimageio CONFIG REQUIRED)
```

I set CMAKE_TOOLCHAIN_FILE in my `CMakeUserPresets.json` like so:

```
"cacheVariables": {
   ...
   "CMAKE_TOOLCHAIN_FILE": "/Users/mark/Projects/github/vcpkg/scripts/buildsystems/vcpkg.cmake",
   ...
}
```

Since CMake was finding the MacPorts version, I added the line

```
    "CMAKE_PREFIX_PATH": "/Users/mark/Projects/github/vcpkg/installed/x64-osx",
```

to try to force it to look in the vcpkg packages first but it doesn't help and in any case I do not like having the architecture explicitly in the path. The same thing happens regardless of whether or not `CONFIG` is present. If I change to `MODULE` it tells me Findopenimageio.cmake was not found found along `CMAKE_MODULE_PATH`.

How can I get find_package to look in the vcpkg packages first? I do not want to have to tell uses they have to uninstall other versions of the package.

**Comment by dg0yt** [comment]
> How can I get find_package to look in the vcpkg packages first?

Are you sure it isn't attempted that way, but fails e.g. due to some dependencies?

You may try to `set(CMAKE_FIND_DEBUG_MODE ON)` (and `OFF`) around you `find_package` call for more information.

**Reply by MarkCallow** [reply]
Thanks for the suggestion to use `set(CMAKE_FIND_DEBUG_MODE ON)`. With the help of that I discovered that cmake must be caching and not updating some hidden values. I'm using the CMake GUI.

- When modifying a config that had previously been used to build using the MacPorts version of openimageio, CMake always "finds" the MacPorts version even though `CMAKE_TOOLCHAIN_FILE` is set to that provided by vcpkg.
- When modifying a config that had not previously used any version of OIIO, I got an error that the directory pointed at by `OpenImageIO_INCLUDES`, `/include` could not be found.

Before trying all this I had changed the find_package line to `find_package(openimageio REQUIRED)`. In both cases above the project was successfully configured to use the  vcpkg provided package, if I first deleted CMake's cache. I later discovered that if I use the original find_package request with `CONFIG` in it, then the error mentioned in the second bullet happens even after the cache has been deleted.

The difference in the two cases is that in the first the vcpkg installed `FindOpenImageIO.cmake` is used, in the second `OpenImageIOConfig.cmake`. The latter prefixes all the paths with `${_IMPORT_PREFIX}` so `${_IMPORT_PREFIX}/include`. Clearly `${_IMPORT_PREFIX}` must be the empty string hence`/include` not found. I have no idea where ${_IMPORT_PREFIX} is supposed to be set.

I'm currently having problems linking which I am investigating.  I have one related question. OIIO has many dependencies. Since vcpkg has installed static libraries for OIIO and all its dependencies, I expect I need to list these in the link command. I was expecting `OpenImageIO_LIBRARIES` to include all these libraries but it does not. Do I have to manually figure out what they are and add them to my link command?

**Reply by MarkCallow** [reply]
> 'm currently having problems linking ...

Indeed all the dependencies are missing from the link. Doesn't `find_package` provide the necessary information?

**Reply by dg0yt** [reply]
`find_package` often needs some help from the port maintainers. There are already open issues for openimageio.
